test/libsinsp_e2e/container.cpp

```diff
@@ -0,0 +1,848 @@
+// in forked children instead of exit(), as _exit() skips all those
+// teardown steps, but this test verifies that even if a child calls
+// exit(), that there aren't any conflicts/races in the static
+// globals, etc.
+//
+// It may be the case someday that there are globals that we don't
+// control or have to keep global that cause conflicts on duplicate
+// exit(), in which case this test will start
+// hanging/failing/crashing. If this happens, we should remove this
+// test.
+
+TEST_F(sys_call_test, container_docker_fork)
+{
+	bool fork_after_container_start = true;
+
+	run_container_docker_test(fork_after_container_start);
+}
+
+TEST_F(sys_call_test, container_docker_bad_socket)
+{
+	bool done = false;
+
+	if (!dutils_check_docker())
+	{
+		printf("Docker not running, skipping test\n");
+		return;
+	}
+
+	event_filter_t filter = [&](sinsp_evt* evt)
+	{
+		if (evt->get_type() == PPME_CONTAINER_JSON_E || evt->get_type() == PPME_CONTAINER_JSON_2_E)
+		{
+			return true;
+		}
+		auto tinfo = evt->get_thread_info();
+		if (tinfo)
+		{
+			return !tinfo->m_container_id.empty();
+		}
+		return false;
+	};
+
+	run_callback_t test = [&](concurrent_object_handle<sinsp> inspector)
+	{
+		ASSERT_TRUE(system("docker kill libsinsp_docker > /dev/null 2>&1 || true") == 0);
+		ASSERT_TRUE(system("docker rm -v libsinsp_docker > /dev/null 2>&1 || true") == 0);
+
+#ifdef __s390x__
+		if (system("docker run -d --name libsinsp_docker s390x/busybox sh -c 'while true; do "
+		           "sleep 1; done'") != 0)
+#else
+		if (system("docker run -d --name libsinsp_docker busybox sh -c 'while true; do sleep 1; "
+		           "done'") != 0)
+#endif
+		{
+			ASSERT_TRUE(false);
+		}
+
+		sleep(2);
+
+		ASSERT_TRUE(system("docker kill libsinsp_docker > /dev/null 2>&1 || true") == 0);
+		ASSERT_TRUE(system("docker rm -v libsinsp_docker > /dev/null 2>&1") == 0);
+	};
+
+	captured_event_callback_t callback = [&](const callback_param& param)
+	{
+		// can't get a container event for failed lookup
+		ASSERT_NE(PPME_CONTAINER_JSON_E, param.m_evt->get_type());
+		ASSERT_NE(PPME_CONTAINER_JSON_2_E, param.m_evt->get_type());
+
+		sinsp_threadinfo* tinfo = param.m_evt->get_thread_info(false);
+		ASSERT_TRUE(tinfo->m_container_id.length() == 12);
+		ASSERT_TRUE(param.m_inspector->m_container_manager.container_exists(tinfo->m_container_id));
+		const auto container_info =
+		    param.m_inspector->m_container_manager.get_container(tinfo->m_container_id);
+		if (container_info && container_info->m_type == CT_DOCKER)
+		{
+			EXPECT_EQ(sinsp_container_lookup::state::FAILED, container_info->get_lookup_status());
+			done = true;
+		}
+	};
+
+	before_open_t setup = [&](sinsp* inspector)
+	{
+		inspector->set_docker_socket_path("/invalid/path");
+	};
+
+	before_close_t cleanup = [&](sinsp* inspector)
+	{ inspector->set_docker_socket_path("/var/run/docker.sock"); };
+
+	ASSERT_NO_FATAL_FAILURE({ event_capture::run(test, callback, filter, setup, cleanup); });
```

Comment:
Also in this test `container_docker_bad_socket` could we at least assert that the socket is bad and what else would make sense to assert (for example assert that `container_info->m_image` is empty etc)? It seems like the exact same assertions as for the successful docker socket, which is what we want to assert 100% as well.